### ISSUE TITLE
Byte code diverges rather than raising Undefined_recursive_module

### ISSUE THREAD

**Issue by arlencox** [ocaml/ocaml#9494]
The following program exhibits different behavior with the byte code compiler than with the native code compiler in OCaml 4.09.0 and 4.10.0:
```ml
(* loop_crash.ml *)
module rec Id : sig
  type t = {id : int}
  val compare : t -> t -> int
end = Id (* error here: undefined compare function *)

module IdSet = Set.Make(Id)

let _ =
  let basic_set = IdSet.singleton {id = 0} in
  IdSet.mem {id = 1 } basic_set (* diverge here *)
```

If I run this program with OCaml `ocaml loop_crash.ml`, the program diverges in the `IdSet.mem` function call.  If I compile the program `ocamlopt loop_crash.ml && ./a.out`, I get the expected `Undefined_recursive_module` exception.

It appears that this problem did not exist in OCaml 4.07.0.

I tested this on both MacOS 10.14.6 and CentOS 7.7.1908.

**Comment by greedy** [comment]
I think the problem is the [change to CamlinternalMod.update_mod](https://github.com/ocaml/ocaml/pull/2279/commits/501bd0bb36034ef62d2a5c74a5e3fd9831f8cc4e#diff-74a6d7dd895b716d1d073887f0e77ae8) in #2279 (is incorrect for the case o == n. In this case probably no action should be taken.

As a side node, it also seems like the optimization that was removed could be retained in the case where the old and new block sizes are the same.

**Comment by xavierleroy** [comment]
Nasty bug.  I agree with @greedy's analysis.  To be fixed along these lines urgently.

**Comment by dra27** [comment]
Here is the repro converted to an ocamltest, to help the fixer:

```diff
diff --git a/testsuite/tests/typing-recmod/loop_crash.ml b/testsuite/tests/typing-recmod/loop_crash.ml
new file mode 100644
index 000000000..12063e1fd
--- /dev/null
+++ b/testsuite/tests/typing-recmod/loop_crash.ml
@@ -0,0 +1,17 @@
+(* TEST
+*)
+
+(* PR#9494 *)
+
+(* loop_crash.ml *)
+module rec Id : sig
+  type t = {id : int}
+  val compare : t -> t -> int
+end = Id (* error here: undefined compare function *)
+
+module IdSet = Set.Make(Id)
+
+let _ = try
+  let basic_set = IdSet.singleton {id = 0} in
+  IdSet.mem {id = 1} basic_set (* diverge here *)
+with e -> print_endline @@ Printexc.to_string e; false
diff --git a/testsuite/tests/typing-recmod/loop_crash.reference b/testsuite/tests/typing-recmod/loop_crash.reference
new file mode 100644
index 000000000..8cabf9f42
--- /dev/null
+++ b/testsuite/tests/typing-recmod/loop_crash.reference
@@ -0,0 +1 @@
+File "loop_crash.ml", line 10, characters 6-12: Undefined recursive module
```

**Comment by dra27** [comment]
I can confirm that that test passes on 4.08 but bytecode hangs on 4.09+

**Comment by mshinwell** [comment]
I will look at this today.  (@xavierleroy Please confirm you haven't already started...)

**Comment by mshinwell** [comment]
It turns out that @lthls has already thought about this, last night, and has some more insights that would be useful to document along with the fix.  We should be hearing from him shortly (he will provide the fix too, to be clear).

**Comment by lthls** [comment]
(Warning: this is quite a long explanation, with a fair number of gory details. The proposed fix is at the end.)

### The core of the issue

Compilation of functions that are exported elements of a safe recursive module definition is done by first creating a function that just raises `Undefined_recursive_module` when called, then later when the module definition has been evaluated fully, patching back the first definition to replace it with the new one, as read from the new module (using `Obj`, of course). But since the new function's closure may be bigger than the initial one, there is actually an intermediate step that first eta-expands the new function (allocating a new closure that is known to be small), before overwriting.
The difference between native and bytecode comes from the fact that in native mode, if we detect that the new closure fits inside the old one, we skip the eta expansion, while in bytecode it is unsafe to do so (if the new closure is strictly smaller, and is the result of a partial application, then the interpreter would assume the wrong number of arguments have been already passed).

This particular bug comes because while the new function is guaranteed, before the overwrite, to behave in a normal manner (raising `Undefined_recursive_module` if it tries to call back into the old version of itself), after the overwrite if it has been eta-expanded and the new function happened to be or call the old one, then calling it calls the eta-expanded version indefinitely.

### The fix
What I propose is to use the non-expanding mode when the sizes of the old and new closures match, even in bytecode (I believe that is the change @xavierleroy was talking about). It is safe with respect to the bytecode interpreter, because we're not cheating with sizes, but the question is: will it prevent the introduction of infinite loops ?

I believe the answer is yes. Because I see only two ways for the eta-expanded version to infinitely call itself:
- either the defining expression evaluates to a regular function, whose body calls back into our original function; but this is a normal infinite loop, and there's no reason why it should raise `Undefined_recursive_module` (and it won't do so on any backend)
- or the defining expression evaluates to itself or another member of the recursive module definition; in this case, the resulting closure will have the same size as the original one (as all initial values for functions have the same size), and no eta-expansion will occur.

Here are some examples slightly more tricky than the first one which show the two cases:

```ocaml
(* Looping version *)
module rec M : sig
  val f : unit -> unit
  val g : unit -> unit
end = struct
  let f = M.g
  let g () = M.f ()
end
```

```ocaml
(* Alias chain *)
module rec M : sig
  val f : unit -> unit
  val g : unit -> unit
end = struct
  let f = M.g
  let g = M.f
end
```

I'm not completely sure that I haven't missed something, because the corresponding code is quite tricky (what happens if we combine this with `let rec` definitions, which can also allocate and patch but with different invariants ? I think that the `let rec` code will never end up patching one of its preallocated blocks with a closure created by `CamlinternalMod.init_mod`, but it's not obvious).

I'm also not convinced that looping instead of raising `Undefined_recursive_module` here is a bug worth fixing. Recursive modules are already hard to specify even when they work, I don't see much value in making them predictable when they don't work. But in case someone wants to submit the fix, here it is:
```diff
diff --git a/stdlib/camlinternalMod.ml b/stdlib/camlinternalMod.ml
index c56d22ef18..9ffe55341d 100644
--- a/stdlib/camlinternalMod.ml
+++ b/stdlib/camlinternalMod.ml
@@ -54,9 +54,10 @@ let rec update_mod shape o n =
       (* The optimisation below is invalid on bytecode since
          the RESTART instruction checks the length of closures.
          See PR#4008 *)
-      if Sys.backend_type = Sys.Native
-      && Obj.tag n = Obj.closure_tag
-      && Obj.size n <= Obj.size o
+      if Obj.tag n = Obj.closure_tag
+      && (Obj.size n = Obj.size o
+          || (Sys.backend_type = Sys.Native
+              && Obj.size n <= Obj.size o))
       then begin overwrite o n end
       else overwrite o (Obj.repr (fun x -> (Obj.obj n : _ -> _) x))
   | Lazy ->
```

**Comment by xavierleroy** [comment]
Thank you for the extensive analysis and the proposed fix.  This is indeed what I had in mind after reading @greedy 's comment.  

It would be a pity to not apply the proposed fix: even if recursive modules are underspecified in many places, consistent behavior between bytecode and native is always preferable, and infinite loops are unpleasant to debug.

**Comment by mshinwell** [comment]
Maybe we could apply the fix, with @dra27 's original testsuite patch, plus @lthls 's extra examples?

**Comment by xavierleroy** [comment]
> Maybe we could apply the fix, with @dra27 's original testsuite patch, plus @lthls 's extra examples?

That would be great.

**Comment by greedy** [comment]
> It would be a pity to not apply the proposed fix: even if recursive modules are underspecified in many places, consistent behavior between bytecode and native is always preferable, and infinite loops are unpleasant to debug.

Very much agreed. It would be a shame to miss reporting a useful error at least in the specific case for o == n because it ought to be a very easy case for the compiler/runtime to detect. The situation arises more easily than it might seem at first.

You might have a module in a recursive set that originally only declared types so it is okay to have `module rec M : sig type t (* = ... *) end = M`. Then during development you add a value to the signature (perhaps indirectly through a ppx) and are interrupted before updating the definition. Now you have code that compiles but produces this hard to debug behavior.

**Comment by arlencox** [comment]
It's worth mentioning that this arose because I had written some code like this:
```ml
module M : sig
  type t = int
end = M
```

Then I realized that I needed M to be comparable, so I added the appropriate deriving line:
```ml
module M : sig
  type t = int
  [@@deriving ord]
end = M
```

Since it compiled I had assumed this was a valid change (having not read the section in the manual that talks about how recursive module initialization works).  Several changes later, I reran the code and was mystified when code that was previously working suddenly no longer worked.

I also think this should be fixed.  I frequently switch between native and byte mode to aid with debugging and any place where the behavior is different certainly problematic.
